impl/src/planning/subplanners.rs

```diff
@@ -636,23 +636,77 @@ impl<'planner> CrateSubplanner<'planner> {
   }
 
   /// Generates source details for internal crate.
-  fn produce_source_details(&self, package: &Package, package_root: &Path) -> SourceDetails {
-    SourceDetails {
-      git_data: self.source_id.as_ref().filter(|id| id.is_git()).map(|id| {
+  fn produce_source_details(
+    &self,
+    crates_io_template: &str,
+    package: &Package,
+    package_root: &Path,
+  ) -> Result<SourceDetails> {
+    let mut git_data = None;
+    let mut download_url = None;
+
+    if let Some(source_id) = self.source_id {
+      if source_id.is_git() {
         let manifest_parent = package.manifest_path.parent().unwrap();
         let path_to_crate_root = manifest_parent.strip_prefix(package_root).unwrap();
         let path_to_crate_root = if path_to_crate_root.components().next().is_some() {
           Some(path_to_crate_root.to_string())
         } else {
           None
         };
-        GitRepo {
-          remote: id.url().to_string(),
-          commit: id.precise().unwrap().to_owned(),
+        git_data = Some(GitRepo {
+          remote: source_id.url().to_string(),
+          commit: source_id.precise().unwrap().to_owned(),
           path_to_crate_root,
+        });
+      }
+      if source_id.is_remote_registry() {
+        download_url = Some(
+          Self::produce_download_url(crates_io_template, package).with_context(|| {
+            format!(
+              "Producing download URL for crate {} version {}",
+              &package.name, &package.version
+            )
+          })?,
+        );
+      }
+    }
+    Ok(SourceDetails {
+      git_data,
+      download_url,
+    })
+  }
+
+  fn produce_download_url(crates_io_template: &str, package: &Package) -> Result<Url> {
+    if let Some(source) = &package.source {
+      if source.is_crates_io() {
+        return Ok(
+          crates_io_template
+            .replace("{crate}", &package.name)
+            .replace("{version}", &package.version.to_string())
+            .parse()?,
+        );
+      }
+      if let Some(index_url) = source.repr.strip_prefix("registry+") {
+        let index = crates_index::BareIndex::from_url(index_url)?;
+        let index = index.open_or_clone()?;
+        let config = index.index_config()?;
+        let url = config.download_url(&package.name, &package.version.to_string());
+        if let Some(url) = url {
+          return Ok(url.parse()?);
```

Comment:
Isn't this already returning a `Result`?
```suggestion
          return url.parse();
```

Reply:
Ish - the `?` is doing an `Into` conversion on the error type. But I've changed this to actually do some wrapping of error messages :)